feat: read Salt-Cloud Maps from the Pillar

nickgarber opened the pull request:
### What does this PR do?
Enable use of cloud map data sourced from the Pillar.
Uses the map data found in the pillar at `cloud:maps:` + the key-name provided with `map_pillar` argument.

Example
```bash
salt-call cloud.map_run map_pillar=my-prd-map
```

Where
```yaml
cloud:
  maps:
    my-prd-map:
      named-profile-01:
        - named-instance-01
        - named-instance-02
```

To preserve backwards compatibility the `map_pillar` option is ignored in the presence of either `map_data` or `path` arguments.

The implementation is in two parts: (1) add maps data from the pillar at `CloudClient.__init__()`  (alongside the existing similarly loaded `profiles` and `providers` info), and (2) allow for selecting and reading from this data in `Map.read()`, which is the function that map data is sourced from.

### What issues does this PR fix or reference?
#40975 

### Previous Behavior
Allow loading map data from a file or from a JSON data structure via CLI.

### New Behavior
Add a pillar-based option to the existing map data.
 
### Tests written?
No so far, but what sort of tests should this have?

### Commits signed with GPG?
No

gtmanfred commented on salt/modules/cloud.py:
```diff
@@ -210,6 +210,8 @@ def map_run(path=None, **kwargs):
 
         salt minionname cloud.map_run /path/to/cloud.map
         salt minionname cloud.map_run map_data='<actual map data>'
+        # to use map data in pillar under salt:cloud:maps:<map_pillar>'
```
I believe this should be `cloud:maps` not `salt:cloud:maps`

nickgarber replied:
Heya @gtmanfred, you're absolutely right! I was referencing it's location in the salt-formula pillar by accident. I have the correction for this on my laptop already - will push ASAP.

gtmanfred commented on salt/modules/cloud.py:
```diff
@@ -210,6 +210,8 @@ def map_run(path=None, **kwargs):
         salt minionname cloud.map_run /path/to/cloud.map
         salt minionname cloud.map_run map_data='<actual map data>'
+        # to use map data in pillar under salt:cloud:maps:<map_pillar>'
+        salt minionname cloud.map_run map_pillar='<map_pillar>'
```
Can you add a section above this for map_pillar and really describe what it is and how to use it?

gtmanfred reviewed (changes requested):
This looks great!

Just a couple docs issue that need to be cleaned up.

cachedout reviewed (changes requested):
Once we get docs pushed that will be reflected here, I think this looks good:

https://docs.saltstack.com/en/latest/topics/cloud/config.html#pillar-configuration

nickgarber committed:
Issue #40975

 - The function signature `def map_run(path=None, **kwargs):`, so
   doesn't need to change.
 - Adding support for Cloud Maps in Pillar builds on the strength of the
   dynamic pillar-centric data organizing scheme.
 - in `Map.__init__`, a file path is read in as the class's
   `self.rendered_map` by the `Map.read` function, (which locates the file
   using the file path that the CloudClient is where most of the
   functionality is collected and it gets a copy of pillar values
   here (https://github.com/saltstack/salt/blob/develop/salt/cloud/init.py#L194-L206).

(doc)
 - added usage and precedence info to `map_run()` at modules/cloud.py
 - added example pillar to topics/cloud/config.rst
 - corrected YAML indentation of example 7 in topics/cloud/map.rst

nickgarber commented:
I've added docs and am just watching for tests. Does this change need to include tests? Any other feedback to consider?

gtmanfred commented:
Tests would be great.

You can put them in unit tests, or if you want to write an integration test, we put them in salt/integration/cloud, and add the @expensiveTest decorator to them.

Thanks!
Daniel

nickgarber committed:
Merge branch 'develop' into develop__wip-issue-40975-read-salt-cloud-maps-from-the-pillar
